### ISSUE TITLE
Upgrade to v4.5.0.01-0 fails

### ISSUE THREAD

**Issue by mtompkins** [jimsihk/moodle-php-nginx#402]
Upgrade to 405 fails.
Logs:

```
Setting up /etc/nginx/nginx.conf...
Setting up /etc/php/conf.d/custom.ini...
Setting up /etc/php/conf.d/custom-opcache-jit.ini...
Setting up /etc/php/php-fpm.d/www.conf...
Starting startup scripts in /docker-entrypoint-init.d ...

*** Running: /docker-entrypoint-init.d/01-uname.sh

Linux moodle 6.6.56-1-lts #1 SMP PREEMPT_DYNAMIC Thu, 10 Oct 2024 12:04:53 +0000 x86_64 Linux

*** Running: /docker-entrypoint-init.d/02-configure-moodle.sh

Waiting for mariadb:3306 to be ready
Z   
11.5.2-MariaDB-ubu2404 �  &fQC<C{d ��- ��      =   cDha-XU%SS*t mysql_native_password mariadb is ready
Using default file session store
Checking maintenance status...
Maintenance mode will be kept enabled
Upgrading moodle...

== Maintenance mode (https://edu.xxxxxxxxxx) ==

Your site is currently in CLI maintenance mode, no web access is allowed.
Checking moodle code version...
From https://github.com/moodle/moodle
 * branch                MOODLE_405_STABLE -> FETCH_HEAD
error: Your local changes to the following files would be overwritten by checkout:
	lib/classes/check/environment/publicpaths.php
Please commit your changes or stash them before you switch branches.

Aborting

*** Failed with return value: 1

Setting up /etc/nginx/nginx.conf...
Setting up /etc/php/conf.d/custom.ini...
Setting up /etc/php/conf.d/custom-opcache-jit.ini...
Setting up /etc/php/php-fpm.d/www.conf...
Starting startup scripts in /docker-entrypoint-init.d ...
```

Loop

I tried overriding the entrypoint and doing a `git stash` in `/var/www/html`
It did not resolve the issue.

**Comment by jimsihk** [comment]
Hi, in your `/var/www/html` directory, is it a separated storage that already with a copy of 404 Moodle?

BTW, you may want to mask some parts of the log in case of server info.

**Comment by mtompkins** [comment]
Thanks for the mention of redacting a bit of the logs. Updated.

> Hi, in your /var/www/html directory, is it a separated storage that already with a copy of 404 Moodle?

It is a mounted volume with the previous version (if I understand your question correctly).

**Comment by jimsihk** [comment]
Then I believe the Moodle 404 also has plugin installed? Are they installed via the web interface or installed via scripts (e.g. the `install-plugin-list` provided in the image)?

If via scripts, one quick workaround could be preparing a new volume to store only the 405 version and install plugin on top of that, then switch to this new container. 

What I am doing is build a new image with the list of plugin downloaded at build time. Then using mounted volume for the `/var/www/moodledata` _(Updated)_  folder only. And disallow plugin installation via the web interface to ensure no missing plugins in the image.

**Comment by mtompkins** [comment]
So the mounts I currently have:
```
    volumes:
      - /etc/localtime:/etc/localtime:ro # Synchronize time of container with the host system
      - /opt/moodle/moodledata:/var/www/moodledata
      - /opt/moodle/moodlehtml:/var/www/html
```
I do not have any plugins installed that aren't defaults (I believe?). If they are installed I would presume by web interface as I have no env variable or any reference to the `install-plugin-list`

Do I need to alter my volume mount from `/var/www/moodledata` to `/var/www/data`?
Perhaps that changed during an upgrade that I missed.

Thanks for your attention as always :)

**Comment by jimsihk** [comment]
Sorry typo, the folder is still `/var/www/moodledata`.

The problem is caused by the modification of the `publicpaths.php` file at [02-configure-moodle.sh#L252](https://github.com/jimsihk/alpine-moodle/blob/405.001.0/rootfs/docker-entrypoint-init.d/02-configure-moodle.sh#L252) which Git will treat as uncommitted change when switching to the 405 branch.

I am developing a hot fix for this scenario.

**Comment by mtompkins** [comment]
Many thanks!

**Comment by mtompkins** [comment]
So what I've come to realize is the purpose of mounting `/var/www/html` is to preserve theme edits (and why "plugins" didn't register as something I was doing). 

I had installed a theme and lost it during this upgrade (no real casualty). A documented way on handling this might be something to consider to prevent others from experiencing it perhaps?.

Sharing the logs hoping it helps. Steps that caused this
- I removed the html mount
- upgraded
- disabled maintenance
- had default theme (makes sense, my theme isn't mounted)
- shutdown
- added the mount back
- startup to error below

```
== Maintenance mode (https://edu.xxxxxxxxxxxx.xxx) ==

Your site is currently in CLI maintenance mode, no web access is allowed.

Checking moodle code version...

fatal: Unable to create '/var/www/html/.git/shallow.lock': File exists.

Another git process seems to be running in this repository, e.g.

an editor opened by 'git commit'. Please make sure all processes

are terminated then try again. If it still fails, a git process

may have crashed in this repository earlier:

remove the file manually to continue.

ERROR!!!  The code you are using is OLDER than the version that made these databases!

*** Failed with return value: 1
```

**Comment by jimsihk** [comment]
Glad to hear that you have found a workaround and luckily the impact is not huge. Definity it is a good idea to have some documentation regarding persistence and upgrade.

Even better, the fix is ready and will be available at the next v4.5.0.01-1 in a couple of hours.

But a manual check on the `/lib/classes/check/environment/publicpaths.php` might be required to revert the previous changes due to a bug at the start up script that it will keep appending `. ":8080"` to some of the lines which the quick fix will not work (yet).

- Keep running below command until there is no more "8080" found in the file.
```
sed -i 's/wwwroot\ \. \"\:8080\"/wwwroot/g' /var/www/html/lib/classes/check/environment/publicpaths.php
```
- Then restart the pod with the fixed version should work.

The latest error (other than the Git temp file) mentioned is thrown by Moodle that the code base (of the mounted disk) is older than the image provided / downloaded. For Moodle minor version upgrades (e.g. 404 to 405), it cannot directly remount the old disk (with 404 in this case) but need to build based on the newer code base (405 in this case); otherwise the database might also reject the connection.
